### DISCUSSION TITLE
关于tun2socks集成Android相关问题

### DISCUSSION THREAD

**Discussion by String-Obj** [xjasonlyu/tun2socks#142] [Q&A]
首先非常感谢作者，这真是一个很厉害的项目。

目前已经集成Android并实现正常代理转发等

有几个疑问想请教作者

如果在Android中不设置“本应用不使用VPN（addDisallowedApplication）”会死循环（但有些需求需要本应用使用VPN）。

Socks5我已经实现UDP转发功能，我会看到很多DNS解析的数据包。

我想在Socks5端接收数据的同时知道域名，但是目前全部是IP。

我是否需要在本地创建一个虚拟DNS的端口用于控制域名呢？

**Comment by xjasonlyu** [comment]
Hi，我对安卓开发了解不多，所以那些选项我不清楚。

DNS这块你需要一个类似fake ip的机制才能获取到对应的域名。

**Comment by String-Obj** [comment]
首先感谢您的回答。

的确，我了解了下您说的fake ip，它可以解决我获取域名的问题，非常感谢你。

我想把这个讨论一直保存，这样其他人集成Android便可以在这里直接找到答案。

**Reply by Hareis** [reply]
你好 我也是被同样的问题所困扰，
adb 输出日志是   I  time="2023-10-28T01:52:09Z" level=info msg="[STACK] fd://84 <-> socks5:///path/to/unix/domain/socket"
然后不知道后续的应该如何操作

**Comment by String-Obj** [comment]
还有个问题，如果我是透明代理并启用了socks5在本机，那么如何解决循环问题呢？

**Reply by xjasonlyu** [reply]
透明代理似乎不需要解决循环问题吧。

**Reply by String-Obj** [reply]
比如：

浏览器发起请求 -> tun2socks  -> socks5 -> 代理请求 -> (这时tun2socks又被截获造成循环)

**Reply by xjasonlyu** [reply]
这不是透明代理吧。

**Reply by String-Obj** [reply]
那可能是因为我刚刚接触导致理解错了。想请教一下您知道如何解决上面的那个问题吗？

**Comment by snltty** [comment]
老哥咋实现的集成Android，自己编译的吗，还是用作者编译的版本

**Comment by lt-123** [comment]
死循环问题，可以在路由列表中排除掉代理服务器的ip，这样就不用DisallowedApplication自己了，新版本android上有这个方法，旧版本需要自己计算路由，排除掉ip
